### PR TITLE
Add autoplace option

### PR DESCRIPTION
**Pull request by huwshimi**
Add controls for auto placing units on the deploy summary.

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3543/
Test FAILed.

**Comment by huwshimi** [comment]
QA:
- add a service
- click deploy
- click 'Automatically place'
- click deploy
- the unit should be placed upon commit
- press shift+?
- tick 'Default to automatically place units on commit'
- click save
- wait for the gui to reload
- add a service
- click deploy
- 'Automatically place' should be the selected option

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3544/
Test FAILed.

File: jujugui/static/gui/src/app/assets/css/base.scss [hunk 1]
```diff
@@ -360,89 +340,6 @@ th {
         }
     }
 }
-/*
- * Intro text that appears on the canvas if there are no deployed services.
- */
-.environment-help {
```
Comment (huwshimi) [inline jujugui/static/gui/src/app/assets/css/base.scss]: Drive-by to move block to own file.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3545/
Test PASSed.

File: jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js [hunk 2]
```diff
@@ -23,6 +23,57 @@ YUI.add('deployment-summary', function() {
...
+      A component for the placement form.
+
+      @method _generatePlacementControl
+      @returns {Object} The placement component.
+    */
+    _generatePlacementControl: function() {
+      var unplacedCount = this.props.getUnplacedUnitCount();
+      var plural = unplacedCount === 1 ? '' : 's';
+      if (unplacedCount === 0) {
+        return;
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js]: So I guess we can define plural later [shrug].

File: jujugui/static/gui/src/app/components/deployment/deployment.js [hunk 3]
```diff
@@ -40,15 +42,19 @@ YUI.add('deployment-component', function() {
 
       @method generateState
       @param {Object} nextProps The props which were sent to the component.
+      @param {Object} state The existing create state.
```
Comment (frankban) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: The existing create state? What does it mean?

File: jujugui/static/gui/src/app/views/utils.js [hunk 4]
```diff
@@ -1697,6 +1696,17 @@ YUI.add('juju-view-utils', function(Y) {
   };
 
   /**
+    Calculate the number of unplaced units.
+
+    @method getUnplacedUnitCount
+    @param {Array} units An array of units.
+    @returns {Object} The number of unplaced units.
```
Comment (frankban) [inline jujugui/static/gui/src/app/views/utils.js]: @returns {number} ?

File: jujugui/static/gui/src/test/test_utils.js [hunk 5]
```diff
@@ -161,6 +161,22 @@ describe('utilities', function() {
     });
   });
 
+  it('can calculate the number of unplaced units', function() {
```
Comment (frankban) [inline jujugui/static/gui/src/test/test_utils.js]: Could you please also test this with 0 units?

**Comment by frankban** [comment]
:+1: with some minor comments.
QA ok, even if using both the session storage and the preference can be weird, e.g. if I set autoplace to True, then commit without autoplace once, at this point the option is no longer honored on subsequent commits.

File: jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js [hunk 6]
```diff
@@ -23,6 +23,57 @@ YUI.add('deployment-summary', function() {
   juju.components.DeploymentSummary = React.createClass({
 
     /**
+      A component for the placement form.
+
+      @method _generatePlacementControl
+      @returns {Object} The placement component.
+    */
+    _generatePlacementControl: function() {
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js]: This feels like a good candidate for the stateless component, thoughts?

File: jujugui/static/gui/src/app/components/deployment/deployment.js [hunk 7]
```diff
@@ -40,15 +42,19 @@ YUI.add('deployment-component', function() {
...
-      };
+    generateState: function(nextProps, state) {
+      if (!state) {
+        state = {};
+      }
+      state.activeComponent = nextProps.activeComponent || 'deployment-bar';
       var hasCommits = this.state ? this.state.hasCommits : false;
       var currentChangeSet = nextProps.currentChangeSet ||
           this.props.currentChangeSet;
+      var autoPlace = state.autoPlace !== undefined ?
```
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: `var autoPlace = state.autoPlace || this.state.autoPlace;`

**Comment by hatched** [comment]
QA Issue (probably pre-existing): 
Add a service to the canvas and switch to the machine view.
Click to deploy and autoplace.
The unit doesn't automatically get placed in the machine view.
Closing and opening the MV places the unit.

**Comment by hatched** [comment]
:+1: QA :ok:
The QA issue above is likely a symptom of a bug already on the board.

File: jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js (hunk 2 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment-summary/deployment-summary.js]: Good call. Done.

File: jujugui/static/gui/src/app/components/deployment/deployment.js (hunk 3 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: Typo, thanks.

File: jujugui/static/gui/src/app/views/utils.js (hunk 4 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/views/utils.js]: Oops, fixed.

File: jujugui/static/gui/src/test/test_utils.js (hunk 5 above)
Comment (huwshimi) [inline jujugui/static/gui/src/test/test_utils.js]: Done.

File: jujugui/static/gui/src/app/components/deployment/deployment.js (hunk 7 above)
Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: This actually needs to check for `undefined` as it needs to retain the value if it is `false`.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3557/
Test FAILed.

File: jujugui/static/gui/src/app/components/deployment/deployment.js (hunk 7 above)
Comment (hatched) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: ahh could you add a comment to that effect or be explicit about checking for `false`?

Comment (huwshimi) [inline jujugui/static/gui/src/app/components/deployment/deployment.js]: Good plan.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3562/
Test PASSed.

**Comment by huwshimi** [comment]
@frankban Yeah, it is a little tricky, it doesn't save your last used choice, it just uses the default each time.

@hatched I've card added for that existing issue.

:shipit:

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/3563/
Test PASSed.

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge
